cocos/animation/core/shared-stack-based-allocator.ts

```diff
@@ -0,0 +1,299 @@
+        const {
+            _sliceSize: sliceLength,
+            _slices: slices,
+            _slicesPerPage: slicesPerPage,
+        } = this;
+
+        const desiredSliceIndex = slices.length;
+
+        let newSliceIndexInPage = 0;
+        let newSlicePageIndex = 0;
+
+        // Specially handle 0 slice length.
+        if (sliceLength === 0) {
+            // If the slice length is 0, we ensure this allocator has and has only one page.
+            // All slices use this page then.
+            if (this[allocatorPageCountTag] === 0) {
+                this._resource.pushPage(this);
+            }
+            assertIsTrue(this[allocatorPageCountTag] === 1);
+        } else {
+            const capacity = slicesPerPage * this[allocatorPageCountTag];
+            assertIsTrue(desiredSliceIndex <= capacity);
+            if (desiredSliceIndex === capacity) {
+                // We need more pages.
+                this._resource.pushPage(this);
+                assertIsTrue(desiredSliceIndex < slicesPerPage * this[allocatorPageCountTag]);
+            }
+
+            newSliceIndexInPage = desiredSliceIndex % slicesPerPage;
+            newSlicePageIndex = (desiredSliceIndex - newSliceIndexInPage) / slicesPerPage;
+            assertIsTrue(this[allocatorPageCountTag] * slicesPerPage >= desiredSliceIndex);
+        }
+
+        const pageMemory = this._resource.getPageMemory(newSlicePageIndex);
+        const newSlice = new SharedMemorySlice(
+            pageMemory,
+            sliceLength * newSliceIndexInPage,
+        );
+        this._slices.push(newSlice);
+        return newSlice;
+    }
+
+    public pop () {
+        const {
+            _slices: slices,
+            _slicesPerPage: slicesPerPage,
+        } = this;
+
+        const allocatedCount = slices.length;
+        assertIsTrue(allocatedCount > 0);
+
+        const removingSliceIndex = allocatedCount - 1;
+        if (this._sliceSize === 0) {
+            // If the slice length is 0, we should pop page if we're popping the last slice,
+            // it's the only page in this allocator.
+            assertIsTrue(this[allocatorPageCountTag] === 1);
+            if (removingSliceIndex === 0) {
+                this._resource.popPage(this);
+            }
+        } else {
+            const removingSliceIndexInPage = removingSliceIndex % slicesPerPage;
+            if (removingSliceIndexInPage === 0) {
+                // Now that the last(beginning) slice of last page is pop-ed,
+                // we pop the last page.
+                this._resource.popPage(this);
+            }
+        }
+
+        this._slices.pop();
+    }
+
+    private _slicesPerPage = 0;
+
+    private _slices: SharedMemorySlice[] = [];
+}
+
+export class SharedStackBasedAllocatorManager {
+    constructor (private _thresholds: number[]) {
+        assertIsTrue(_thresholds.every((v, i, arr) => i === 0 || v > arr[i - 1]));
+    }
+
+    public get empty () {
```

Comment:
`isEmpty` ?

Reply:
If it's `isEmpty`, it should be a function instead of a property right?

Reply:
I think it's ok to use `isEmpty` in a property. why need a function ?

Reply:
Agreed.

Reply:
https://github.com/cocos/cocos-engine/pull/13503/commits/5c4b1ba7c485c466ba071aec373f55e87c813bb5